File: C++/2.两数相加.cpp
/*
 * @lc app=leetcode.cn id=2 lang=cpp
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *head = new ListNode(0);
        ListNode *ptr = head;
        int carry = 0;
        while (true)
        {
            if (l1 != NULL)
            {
                carry += l1->val;
                l1 = l1->next;
            }
            if (l2 != NULL)
            {
                carry += l2->val;
                l2 = l2->next;
            }
            ptr->val = carry % 10;
            carry /= 10;
            if (l1 != NULL || l2 != NULL || carry != 0)
            {
                ptr = (ptr->next = new ListNode(0));
            }
            else
                break;
        }
        return head;
    }
};
// @lc code=end

File: JavaScript/2.两数相加.js
/*
 * @lc app=leetcode.cn id=2 lang=javascript
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  const head = new ListNode(0);
  let ptr = head;
  let carry = 0;

  while (true) {
    if (l1 !== null) {
      carry += l1.val;
      l1 = l1.next;
    }
    if (l2 !== null) {
      carry += l2.val;
      l2 = l2.next;
    }

    ptr.val = carry % 10;
    carry = Math.floor(carry / 10);

    if (l1 !== null || l2 !== null || carry !== 0) {
      ptr.next = new ListNode(0);
      ptr = ptr.next;
    } else {
      break;
    }
  }

  return head;
};
// @lc code=end
